# 动手学数据分析

##  第一章——数据加载任务系列：

### 任务一：导入numpy和pandas

In [2]:
import numpy as np
import pandas as pd

### 任务二：载入数据

In [6]:
df = pd.read_csv('D:\\DataWhale\\Titanic\\train.csv')

In [7]:
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [8]:
df = pd.read_csv('train.csv')
df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [22]:
print('行列数')
df.values.shape

行列数


(891, 12)

In [17]:
df2 = pd.read_table('train.csv',nrows =3)
#nrows 需要读取的行数（从文件头开始算起）。

In [21]:
df2

PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0  1,0,3,"Braund, Mr. Owen Harris",male,22,1,0,A/...                             
1  2,1,1,"Cumings, Mrs. John Bradley (Florence Br...                             
2  3,1,3,"Heikkinen, Miss. Laina",female,26,0,0,S...

In [23]:
print('行列数')
df2.values.shape

行列数


(3, 1)

## pd.read_csv()和pd.read_table()的不同：

read_table(),读完后每个字符串之间有逗号相隔，其实表明每一行作为一个维度进行了存储，所以最后它是一个3行1列的数组，每一行字符串为一列.

read_csv(), 读完后是一个891行12列的数组，每一个字符串作为一列，这是二者的主要区别。

### 任务三：每1000行为一个数据模块，逐块读取

In [41]:
chunk = pd.read_csv('train.csv', chunksize =1000)
#chunk.head(5)

### 【思考】什么是逐块读取？为什么要逐块读取呢？
read_csv中有个参数chunksize，通过指定一个chunksize分块大小来读取文件，返回的是一个可迭代的对象TextFileReader。

方便读取其中的一部分数据或对文件进行逐块处理。

In [42]:
for ck in chunk:
    print(ck.head())

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  


### 任务四：将表头改成中文，索引改为乘客ID 


PassengerId => 乘客ID
Survived => 是否幸存
Pclass => 乘客等级(1/2/3等舱位)
Name => 乘客姓名
Sex => 性别
Age => 年龄
SibSp => 堂兄弟/妹个数
Parch => 父母与小孩个数
Ticket => 船票信息
Fare => 票价
Cabin => 客舱
Embarked => 登船港口

In [47]:
df3 = pd.read_csv('train.csv', names =['乘客ID','是否幸存','仓位等级','姓名','性别','年龄','兄弟姐妹个数','父母子女个数','船票信息','票价','客舱','登船港口'],index_col='乘客ID',header = 0)
# header = 0 用于去掉原来的英文表头.指定行数用来作为列名，数据开始行数。如果文件中没有列名，则默认为0，否则设置为None。如果明确设定header=0 就会替换掉原来存在列名。
# names 用于结果的列名列表，如果数据文件中没有列标题行，就需要执行header=None。
# index_col 用作行索引的列编号或者列名
df3.head() 

是否幸存  仓位等级                                                 姓名      性别  \
乘客ID                                                                          
1        0     3                            Braund, Mr. Owen Harris    male   
2        1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
3        1     3                             Heikkinen, Miss. Laina  female   
4        1     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
5        0     3                           Allen, Mr. William Henry    male   

        年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
乘客ID                                                              
1     22.0       1       0         A/5 21171   7.2500   NaN    S  
2     38.0       1       0          PC 17599  71.2833   C85    C  
3     26.0       0       0  STON/O2. 3101282   7.9250   NaN    S  
4     35.0       1       0            113803  53.1000  C123    S  
5     35.0       0       0            373450   8.0500   NaN    S

【思考】所谓将表头改为中文其中一个思路是：将英文表头替换成中文。还有其他的方法吗？
方法二如下：

In [60]:
df4 = pd.read_csv('train.csv')
df4.columns = ['乘客ID','是否幸存','仓位等级','姓名','性别','年龄','兄弟姐妹个数','父母子女个数','船票信息','票价','客舱','登船港口']
df4 = df4.set_index('乘客ID') # 设置第1列为索引，若去掉则默认自动添加1列行号
df4.head(3)

是否幸存  仓位等级                                                 姓名      性别  \
乘客ID                                                                          
1        0     3                            Braund, Mr. Owen Harris    male   
2        1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
3        1     3                             Heikkinen, Miss. Laina  female   

        年龄  兄弟姐妹个数  父母子女个数              船票信息       票价   客舱 登船港口  
乘客ID                                                             
1     22.0       1       0         A/5 21171   7.2500  NaN    S  
2     38.0       1       0          PC 17599  71.2833  C85    C  
3     26.0       0       0  STON/O2. 3101282   7.9250  NaN    S

## 数据观察任务系列：
导入数据后，你可能要对数据的整体结构和样例进行概览，比如说，数据大小、有多少列，各列都是什么格式的，是否包含null等

### 任务一：查看数据的基本信息

In [61]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 1 to 891
Data columns (total 11 columns):
是否幸存      891 non-null int64
仓位等级      891 non-null int64
姓名        891 non-null object
性别        891 non-null object
年龄        714 non-null float64
兄弟姐妹个数    891 non-null int64
父母子女个数    891 non-null int64
船票信息      891 non-null object
票价        891 non-null float64
客舱        204 non-null object
登船港口      889 non-null object
dtypes: float64(2), int64(4), object(5)
memory usage: 83.5+ KB


In [68]:
df.index

RangeIndex(start=0, stop=891, step=1)

In [66]:
print(df4.index)

Int64Index([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,
            ...
            882, 883, 884, 885, 886, 887, 888, 889, 890, 891],
           dtype='int64', name='乘客ID', length=891)


In [69]:
df4.columns #查看表头

Index(['是否幸存', '仓位等级', '姓名', '性别', '年龄', '兄弟姐妹个数', '父母子女个数', '船票信息', '票价',
       '客舱', '登船港口'],
      dtype='object')

In [70]:
df.shape # 查看行列数

(891, 12)

In [71]:
cols = df.columns
for col in cols:
    print(col + ':' + str(df[col].dtype))

PassengerId:int64
Survived:int64
Pclass:int64
Name:object
Sex:object
Age:float64
SibSp:int64
Parch:int64
Ticket:object
Fare:float64
Cabin:object
Embarked:object


In [74]:
df.isna().sum() # 查看缺失值情况

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### 任务二：观察表格前10行的数据和后5行的数据

In [75]:
df4.head(10)

是否幸存  仓位等级                                                 姓名      性别  \
乘客ID                                                                          
1        0     3                            Braund, Mr. Owen Harris    male   
2        1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
3        1     3                             Heikkinen, Miss. Laina  female   
4        1     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
5        0     3                           Allen, Mr. William Henry    male   
6        0     3                                   Moran, Mr. James    male   
7        0     1                            McCarthy, Mr. Timothy J    male   
8        0     3                     Palsson, Master. Gosta Leonard    male   
9        1     3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female   
10       1     2                Nasser, Mrs. Nicholas (Adele Achem)  female   

        年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
乘客ID                                                              
1     22.0       1       0         A/5 21171   7.2500   NaN    S  
2     38.0       1       0          PC 17599  71.2833   C85    C  
3     26.0       0       0  STON/O2. 3101282   7.9250   NaN    S  
4     35.0       1       0            113803  53.1000  C123    S  
5     35.0       0       0            373450   8.0500   NaN    S  
6      NaN       0       0            330877   8.4583   NaN    Q  
7     54.0       0       0             17463  51.8625   E46    S  
8      2.0       3       1            349909  21.0750   NaN    S  
9     27.0       0       2            347742  11.1333   NaN    S  
10    14.0       1       0            237736  30.0708   NaN    C

In [76]:
df.tail(5)

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

### 任务三：判断数据是否为空

In [77]:
df.isnull().head()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0        False     False   False  False  False  False  False  False   False   
1        False     False   False  False  False  False  False  False   False   
2        False     False   False  False  False  False  False  False   False   
3        False     False   False  False  False  False  False  False   False   
4        False     False   False  False  False  False  False  False   False   

    Fare  Cabin  Embarked  
0  False   True     False  
1  False  False     False  
2  False   True     False  
3  False  False     False  
4  False   True     False

上面的操作都是数据分析中对于数据本身的观察.

## 保存数据任务系列

 任务一：将你加载并做出改变的数据，在工作目录下保存为一个新文件

In [79]:
df4.to_csv('train_chinese.csv')
df4.to_csv('D:\\DataWhale\\Titanic\\train_chinese.csv')

【总结】数据的加载以及入门，接下来就要接触数据本身的运算，我们将主要掌握numpy和pandas在工作和项目场景的运用。

## 第一章—— Pandas 基础

### 任务一：pandas中有两个数据类型DateFrame和Series，通过查找简单了解
Series：一种类似于一维数组的对象，它是由一组数据 以及一组与之相关的数据标签（即索引）组成。

DataFrame：一个表格型的数据结构，含有一组有序的列，每列可以是不同的值 类型（数值、字符串、布尔值等），DataFrame既有行索引也有列索引。

In [80]:
import numpy as np
import pandas as pd
print(pd.__version__)

0.23.4


In [85]:
scores = {'语文': 95,'数学': 100,'英语': 99,'体育': 100} 
# 定义字典，然后创建Series， 字典中的key组成Series的索引， 字典中的value组成Series中的values
series01 = pd.Series(scores) # 注意Series是大写字母开头.
series01

语文     95
数学    100
英语     99
体育    100
dtype: int64

In [87]:
series02 = pd.Series([90,97,98,95], index = ['语文','数学','英语','体育'])
series02
# 通过数组创建Series,  指定索引，没有指定的话，默认自动创建0--N-1的整型索引。

语文    90
数学    97
英语    98
体育    95
dtype: int64

In [89]:
series02.index

Index(['语文', '数学', '英语', '体育'], dtype='object')

In [90]:
series02.values

array([90, 97, 98, 95], dtype=int64)

In [91]:
series02['数学']

97

In [92]:
series02[1]

97

In [94]:
series02[series02 > 95]

数学    97
英语    98
dtype: int64

#### Sereies缺失值检测

pandas中的isnull和notnull函数可用于Series缺失值检测

In [96]:
series03 = pd.Series(series02, index =['语文', '数学', '英语', '体育','美术'])
series03

语文    90.0
数学    97.0
英语    98.0
体育    95.0
美术     NaN
dtype: float64

In [97]:
pd.isnull(series03)

语文    False
数学    False
英语    False
体育    False
美术     True
dtype: bool

In [98]:
series03[pd.isnull(series03)]

美术   NaN
dtype: float64

In [99]:
series03[pd.notnull(series03)]

语文    90.0
数学    97.0
英语    98.0
体育    95.0
dtype: float64

In [106]:
#通过字典方式创建DataFrame
data = {'国家':['中国','日本','美国','法国','俄罗斯'],'首都':['北京','东京','华盛顿','巴黎','Null']}
df01 =pd.DataFrame(data)
df01

国家    首都
0   中国    北京
1   日本    东京
2   美国   华盛顿
3   法国    巴黎
4  俄罗斯  Null

In [112]:
#通过二维数组创建DataFrame
arr = np.array([['语文',92],['数学',93],['英语', 94],['体育',np.NaN]])
df02 = pd.DataFrame(arr, columns =['科目','成绩'])
df02

科目   成绩
0  语文   92
1  数学   93
2  英语   94
3  体育  nan

In [113]:
df02.isnull()

科目     成绩
0  False  False
1  False  False
2  False  False
3  False  False

In [114]:
df02.notnull()

科目    成绩
0  True  True
1  True  True
2  True  True
3  True  True

In [116]:
df02.dropna() #丢弃包含缺失值的行

科目   成绩
0  语文   92
1  数学   93
2  英语   94
3  体育  nan

### 任务二： 载入train.csv文件
### 任务三和任务四：  查看 数据列

In [117]:
df = pd.read_csv('train_chinese.csv')
df.head(3)

乘客ID  是否幸存  仓位等级                                                 姓名  \
0     1     0     3                            Braund, Mr. Owen Harris   
1     2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2     3     1     3                             Heikkinen, Miss. Laina   

       性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价   客舱 登船港口  
0    male  22.0       1       0         A/5 21171   7.2500  NaN    S  
1  female  38.0       1       0          PC 17599  71.2833  C85    C  
2  female  26.0       0       0  STON/O2. 3101282   7.9250  NaN    S

In [118]:
df.columns

Index(['乘客ID', '是否幸存', '仓位等级', '姓名', '性别', '年龄', '兄弟姐妹个数', '父母子女个数', '船票信息',
       '票价', '客舱', '登船港口'],
      dtype='object')

In [120]:
df['客舱'].head(3)

0    NaN
1    C85
2    NaN
Name: 客舱, dtype: object

In [122]:
df.客舱.head(3)

0    NaN
1    C85
2    NaN
Name: 客舱, dtype: object

In [124]:
# 以年龄为条件筛选，显示20岁以下的乘客信息
df[df['年龄'] < 20].head()

乘客ID  是否幸存  仓位等级                                    姓名      性别    年龄  \
7      8     0     3        Palsson, Master. Gosta Leonard    male   2.0   
9     10     1     2   Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
10    11     1     3       Sandstrom, Miss. Marguerite Rut  female   4.0   
14    15     0     3  Vestrom, Miss. Hulda Amanda Adolfina  female  14.0   
16    17     0     3                  Rice, Master. Eugene    male   2.0   

    兄弟姐妹个数  父母子女个数     船票信息       票价   客舱 登船港口  
7        3       1   349909  21.0750  NaN    S  
9        1       0   237736  30.0708  NaN    C  
10       1       1  PP 9549  16.7000   G6    S  
14       0       0   350406   7.8542  NaN    S  
16       4       1   382652  29.1250  NaN    Q

In [126]:
youngers = df[(df['年龄'] > 20) & (df['年龄'] < 40)]
youngers.head() 

乘客ID  是否幸存  仓位等级                                                 姓名  \
0     1     0     3                            Braund, Mr. Owen Harris   
1     2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2     3     1     3                             Heikkinen, Miss. Laina   
3     4     1     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4     5     0     3                           Allen, Mr. William Henry   

       性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
0    male  22.0       1       0         A/5 21171   7.2500   NaN    S  
1  female  38.0       1       0          PC 17599  71.2833   C85    C  
2  female  26.0       0       0  STON/O2. 3101282   7.9250   NaN    S  
3  female  35.0       1       0            113803  53.1000  C123    S  
4    male  35.0       0       0            373450   8.0500   NaN    S

### 任务五：加载文件"test_1.csv"，然后对比"train.csv"，看看有哪些多出的列，然后将多出的列删除

In [128]:
test_1 = pd.read_csv('test.csv')
test_1.head(3)

PassengerId  Pclass                              Name     Sex   Age  SibSp  \
0          892       3                  Kelly, Mr. James    male  34.5      0   
1          893       3  Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   
2          894       2         Myles, Mr. Thomas Francis    male  62.0      0   

   Parch  Ticket    Fare Cabin Embarked  
0      0  330911  7.8292   NaN        Q  
1      0  363272  7.0000   NaN        S  
2      0  240276  9.6875   NaN        Q

### 任务六： 将['PassengerId','Name','Age','Ticket']这几个列元素隐藏，只观察其他几个列元素

In [132]:
df05 = pd.read_csv('train.csv')
df05.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [133]:
df05.drop(['PassengerId','Name','Age','Ticket'],axis=1).head(3)

Survived  Pclass     Sex  SibSp  Parch     Fare Cabin Embarked
0         0       3    male      1      0   7.2500   NaN        S
1         1       1  female      1      0  71.2833   C85        C
2         1       3  female      0      0   7.9250   NaN        S

In [134]:
df05.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

【思考】对比任务五和任务六，是不是使用了不一样的方法(函数)，如果使用一样的函数如何完成上面的不同的要求呢？

【思考回答】

如果想要完全的删除你的数据结构，使用inplace=True，因为使用inplace就将原数据覆盖了，所以这里没有用。

## 筛选的逻辑
表格数据中，最重要的一个功能就是要具有可筛选的能力，选出我所需要的信息，丢弃无用的信息。

### 任务一： 我们以"Age"为筛选条件，显示年龄在10岁以下的乘客信息。
### 任务二： 以"Age"为条件，将年龄在10岁以上和50岁以下的乘客信息显示出来，并将这个数据命名为midage
### 任务三：将midage的数据中第100行的"Pclass"和"Sex"的数据显示出来
### 任务四：使用loc方法将midage的数据中第100，105，108行的"Pclass"，"Name"和"Sex"的数据显示出来
### 任务五：使用iloc方法将midage的数据中第100，105，108行的"Pclass"，"Name"和"Sex"的数据显示出来

In [135]:
df05[df05["Age"]<10].head(3)

PassengerId  Survived  Pclass                             Name     Sex  \
7             8         0       3   Palsson, Master. Gosta Leonard    male   
10           11         1       3  Sandstrom, Miss. Marguerite Rut  female   
16           17         0       3             Rice, Master. Eugene    male   

    Age  SibSp  Parch   Ticket    Fare Cabin Embarked  
7   2.0      3      1   349909  21.075   NaN        S  
10  4.0      1      1  PP 9549  16.700    G6        S  
16  2.0      4      1   382652  29.125   NaN        Q

In [136]:
midage = df05[(df05["Age"]>10)& (df05["Age"]<50)]
midage.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [137]:
midage = midage.reset_index(drop=True)
midage.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

【思考】这个reset_index()函数的作用是什么？如果不用这个函数，下面的任务会出现什么情况？

In [138]:
midage.loc[[100],['Pclass','Sex']]

Pclass   Sex
100       2  male

In [139]:
midage.loc[[100,105,108],['Pclass','Name','Sex']] #因为你主动的延长了行的距离，所以会产生表格形式

Pclass                               Name   Sex
100       2  Byles, Rev. Thomas Roussel Davids  male
105       3           Cribb, Mr. John Hatfield  male
108       3                    Calic, Mr. Jovo  male

【提示】使用pandas提出的简单方式，你可以看看loc方法

对比整体的数据位置，你有发现什么问题吗？那么如何解决？

In [140]:
midage.iloc[[100,105,108],[2,3,4]]

Pclass                               Name   Sex
100       2  Byles, Rev. Thomas Roussel Davids  male
105       3           Cribb, Mr. John Hatfield  male
108       3                    Calic, Mr. Jovo  male

## 第一章——探索性数据分析

### 任务一：利用Pandas对示例数据进行排序，要求升序
### 任务二：对泰坦尼克号数据（trian.csv）按票价和年龄两列进行综合排序（降序排列）
###  任务三：利用Pandas进行算术计算，计算两个DataFrame数据相加结果
### 任务四：通过泰坦尼克号数据如何计算出在船上最大的家族有多少人？
### 任务五：学会使用Pandas describe()函数查看数据基本统计信息
### 任务六：分别看看泰坦尼克号数据集中 票价、父母子女 这列数据的基本统计数据


In [141]:
import numpy as np
import pandas as pd

In [142]:
text = pd.read_csv('train_chinese.csv')
text.head()

乘客ID  是否幸存  仓位等级                                                 姓名  \
0     1     0     3                            Braund, Mr. Owen Harris   
1     2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2     3     1     3                             Heikkinen, Miss. Laina   
3     4     1     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4     5     0     3                           Allen, Mr. William Henry   

       性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
0    male  22.0       1       0         A/5 21171   7.2500   NaN    S  
1  female  38.0       1       0          PC 17599  71.2833   C85    C  
2  female  26.0       0       0  STON/O2. 3101282   7.9250   NaN    S  
3  female  35.0       1       0            113803  53.1000  C123    S  
4    male  35.0       0       0            373450   8.0500   NaN    S

教材《Python for Data Analysis》第五章

In [144]:
# 具体请看《利用Python进行数据分析》第五章 排序和排名 部分

#自己构建一个都为数字的DataFrame数据
frame = pd.DataFrame(np.arange(8).reshape((2, 4)), 
                     index=['2', '1'], 
                     columns=['d', 'a', 'b', 'c'])

#pd.DataFrame() ：创建一个DataFrame对象
#np.arange(8).reshape((2, 4)) : 产生8以内的0-7数字,生成一个二维数组（2*4）,第一列：0，1，2，3 第二列：4，5，6，7
#index=['2, 1] ：DataFrame 对象的索引列
#columns=['d', 'a', 'b', 'c'] ：DataFrame 对象的索引行
frame

d  a  b  c
2  0  1  2  3
1  4  5  6  7

In [145]:
# 大多数时候我们都是想根据列的值来排序,所以，将你构建的DataFrame中的数据根据某一列，升序排列 
frame.sort_values(by='c', ascending=False)

#可以看到sort_values这个函数中by参数指向要排列的列，ascending参数指向排序的方式（升序还是降序）

d  a  b  c
1  4  5  6  7
2  0  1  2  3

In [146]:
# 让行索引升序排序
frame.sort_index()

d  a  b  c
1  4  5  6  7
2  0  1  2  3

In [147]:
# 让列索引升序排序
frame.sort_index(axis=1)

a  b  c  d
2  1  2  3  0
1  5  6  7  4

In [148]:
# 让列索引降序排序
frame.sort_index(axis=1, ascending=False)

d  c  b  a
2  0  3  2  1
1  4  7  6  5

In [149]:
# 让任选两列数据同时降序排序
frame.sort_values(by=['a', 'c'])

d  a  b  c
2  0  1  2  3
1  4  5  6  7

In [150]:
'''
在开始我们已经导入了train_chinese.csv数据，而且前面我们也学习了导入数据过程，根据上面学习，我们直接对目标列进行排序即可
head(20) : 读取前20条数据
'''

text.sort_values(by=['票价', '年龄'], ascending=False).head(3)

乘客ID  是否幸存  仓位等级                                  姓名      性别    年龄  \
679   680     1     1  Cardeza, Mr. Thomas Drake Martinez    male  36.0   
258   259     1     1                    Ward, Miss. Anna  female  35.0   
737   738     1     1              Lesurer, Mr. Gustave J    male  35.0   

     兄弟姐妹个数  父母子女个数      船票信息        票价           客舱 登船港口  
679       0       1  PC 17755  512.3292  B51 B53 B55    C  
258       0       0  PC 17755  512.3292          NaN    C  
737       0       0  PC 17755  512.3292         B101    C

【思考】排序后，如果我们仅仅关注年龄和票价两列。根据常识我知道发现票价越高的应该客舱越好，所以我们会明显看出，票价前20的乘客中存活的有14人，这是相当高的一个比例，那么我们后面是不是可以进一步分析一下票价和存活之间的关系，年龄和存活之间的关系呢？当你开始发现数据之间的关系了，数据分析就开始了。

In [151]:
# 具体请看《利用Python进行数据分析》第五章 算术运算与数据对齐 部分

#建立一个例子
frame1_a = pd.DataFrame(np.arange(9.).reshape(3, 3),
                     columns=['a', 'b', 'c'],
                     index=['one', 'two', 'three'])
frame1_b = pd.DataFrame(np.arange(12.).reshape(4, 3),
                     columns=['a', 'e', 'c'],
                     index=['first', 'one', 'two', 'second'])
frame1_a

a    b    c
one    0.0  1.0  2.0
two    3.0  4.0  5.0
three  6.0  7.0  8.0

In [153]:
frame1_b

a     e     c
first   0.0   1.0   2.0
one     3.0   4.0   5.0
two     6.0   7.0   8.0
second  9.0  10.0  11.0

In [154]:
#将frame_a和frame_b进行相加
frame1_a + frame1_b

a   b     c   e
first   NaN NaN   NaN NaN
one     3.0 NaN   7.0 NaN
second  NaN NaN   NaN NaN
three   NaN NaN   NaN NaN
two     9.0 NaN  13.0 NaN

【提醒】两个DataFrame相加后，会返回一个新的DataFrame，对应的行和列的值会相加，没有对应的会变成空值NaN。
当然，DataFrame还有很多算术运算，如减法，除法等，有兴趣的同学可以看《利用Python进行数据分析》第五章 算术运算与数据对齐 部分，多在网络上查找相关学习资料。

In [155]:
'''
还是用之前导入的chinese_train.csv如果我们想看看在船上，最大的家族有多少人（‘兄弟姐妹个数’+‘父母子女个数’），我们该怎么做呢？
'''
max(text['兄弟姐妹个数'] + text['父母子女个数'])

10

In [156]:
#(1) 关键知识点示例做一遍（简单数据）
# 具体请看《利用Python进行数据分析》第五章 汇总和计算描述统计 部分

#建立一个例子
frame2 = pd.DataFrame([[1.4, np.nan], 
                       [7.1, -4.5],
                       [np.nan, np.nan], 
                       [0.75, -1.3]
                      ], index=['a', 'b', 'c', 'd'], columns=['one', 'two'])
frame2

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

In [158]:
# 调用 describe 函数，观察frame2的数据基本信息

frame2.describe()

one       two
count  3.000000  2.000000
mean   3.083333 -2.900000
std    3.493685  2.262742
min    0.750000 -4.500000
25%    1.075000 -3.700000
50%    1.400000 -2.900000
75%    4.250000 -2.100000
max    7.100000 -1.300000

'''
count : 样本数据大小
mean : 样本数据的平均值
std : 样本数据的标准差
min : 样本数据的最小值
25% : 样本数据25%的时候的值
50% : 样本数据50%的时候的值
75% : 样本数据75%的时候的值
max : 样本数据的最大值
'''

In [159]:
'''
看看泰坦尼克号数据集中 票价 这列数据的基本统计数据
'''
text['票价'].describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: 票价, dtype: float64

In [160]:
'''
通过上面的例子，我们再看看泰坦尼克号数据集中 父母子女个数 这列数据的基本统计数据，然后可以说出你的想法
'''
text['父母子女个数'].describe()

count    891.000000
mean       0.381594
std        0.806057
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        6.000000
Name: 父母子女个数, dtype: float64

【总结】本节中我们通过Pandas的一些内置函数对数据进行了初步统计查看，这个过程最重要的不是大家得掌握这些函数，而是看懂从这些函数出来的数据，构建自己的数据分析思维，这也是第一章最重要的点，希望大家学完第一章能对数据有个基本认识，了解自己在做什么，为什么这么做，后面的章节我们将开始对数据进行清洗，进一步分析。